File: scrudbeans-integration-tests/src/test/java/mypackage/test/RestErrorsIT.java
package mypackage.test;

import static io.restassured.RestAssured.given;
import static org.junit.jupiter.api.Assertions.assertEquals;

import java.util.Set;

import com.github.manosbatsis.scrudbeans.api.error.ConstraintViolationEntry;
import com.github.manosbatsis.scrudbeans.test.AbstractRestAssueredIT;
import lombok.extern.slf4j.Slf4j;
import mypackage.ScrudBeansSampleApplication;
import mypackage.model.DiscountCode;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;

import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.test.context.junit.jupiter.SpringExtension;

@Slf4j
@ExtendWith(SpringExtension.class)
@SpringBootTest(classes = ScrudBeansSampleApplication.class, webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
public class RestErrorsIT extends AbstractRestAssueredIT {

	@Test
	public void testNotFound() {
		SimpleErrorResponse error = given()
				.spec(defaultSpec())
				.get("api/rest/products/invalid")
				.then()
				.statusCode(404).extract().as(SimpleErrorResponse.class);

		assertEquals("Not Found", error.getTitle());
		assertEquals(404, error.getHttpStatusCode().intValue());
	}

	@Test
	public void testBadRequestNotNullConstraint() {

		DiscountCode discountCode = new DiscountCode();
		discountCode.setCode("DISCOUNT_1");

		SimpleErrorResponse error = given()
				.spec(defaultSpec())
				.body(discountCode)
				.post("api/rest/discountCodes")
				.then()
				.statusCode(400).extract().as(SimpleErrorResponse.class);

		assertEquals("Validation failed", error.getTitle());
		assertEquals(400, error.getHttpStatusCode().intValue());
		Set<ConstraintViolationEntry> violations = error.getValidationErrors();
		assertEquals(1, violations.size());
		ConstraintViolationEntry violation = violations.iterator().next();
		assertEquals("must not be null", violation.getMessage());
		assertEquals("percentage", violation.getPropertyPath());
	}

	/**
	 * Test meaningful messages in constraints validation
	 */
	@Test
	public void testBadRequestUniqueConstraint() {

		DiscountCode discountCode1 = new DiscountCode();
		discountCode1.setCode("DISCOUNT_1");
		discountCode1.setPercentage(10);

		discountCode1 = given()
				.spec(defaultSpec())
				.body(discountCode1)
				.post("/api/rest/discountCodes")
				.then()
				.statusCode(201).extract().as(DiscountCode.class);

		DiscountCode discountCode2 = new DiscountCode();
		discountCode2.setCode("DISCOUNT_1");
		discountCode2.setPercentage(10);
		SimpleErrorResponse error = given()
				.spec(defaultSpec())
				.body(discountCode2)
				.post("/api/rest/discountCodes")
				.then()
				.statusCode(400).extract().as(SimpleErrorResponse.class);

		assertEquals("Validation failed", error.getTitle());
		assertEquals(400, error.getHttpStatusCode().intValue());
		Set<ConstraintViolationEntry> violations = error.getValidationErrors();
		assertEquals(1, violations.size());
		ConstraintViolationEntry violation = violations.iterator().next();
		assertEquals("Unique value not available for property: code", violation.getMessage());
		assertEquals("code", violation.getPropertyPath());
	}

}

File: scrudbeans-integration-tests-kotlin/src/test/kotlin/mykotlinckage/test/RestErrorsIT.kt
package mykotlinpackage.test

import com.github.manosbatsis.scrudbeans.api.error.ConstraintViolationEntry
import com.github.manosbatsis.scrudbeans.test.AbstractRestAssuredIT
import io.restassured.RestAssured
import mykotlinpackage.ScrudBeansSampleApplication
import mykotlinpackage.model.DiscountCode
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.boot.test.context.SpringBootTest.WebEnvironment.RANDOM_PORT
import org.springframework.boot.test.web.client.TestRestTemplate
import org.springframework.http.HttpStatus
import org.springframework.http.client.BufferingClientHttpRequestFactory
import org.springframework.http.client.SimpleClientHttpRequestFactory
import org.springframework.test.context.junit.jupiter.SpringExtension

@ExtendWith(SpringExtension::class)
@SpringBootTest(classes = [ScrudBeansSampleApplication::class], webEnvironment = RANDOM_PORT)
class RestErrorsIT : AbstractRestAssuredIT() {

    @Autowired
    lateinit var restTemplateOrig: TestRestTemplate

    val restTemplate: TestRestTemplate by lazy {
        restTemplateOrig.restTemplate.setRequestFactory(BufferingClientHttpRequestFactory(SimpleClientHttpRequestFactory()))
        //restTemplateOrig.restTemplate.interceptors.add(RequestResponseLoggingInterceptor())
        restTemplateOrig
    }

    @Test
    fun testNotFound() {
        val error: SimpleErrorResponse = RestAssured.given()
                .spec(defaultSpec())["api/rest/products/invalid"]
                .then()
                .statusCode(404).extract().`as`(SimpleErrorResponse::class.java)
        assertEquals("Not Found", error.message)
        assertEquals(404, error.httpStatusCode)
    }



    @Test
    fun testBadRequestNotNullConstraint() {
        val discountCode = DiscountCode(code = "DISCOUNT_testBadRequestNotNullConstraint")
        val error: SimpleErrorResponse = RestAssured.given()
                .spec(defaultSpec())
                .body(discountCode)
                .post("api/rest/discountCodes")
                .then()
                .statusCode(400).extract().`as`(SimpleErrorResponse::class.java)
        assertEquals("Validation failed", error.message)
        assertEquals(400, error.httpStatusCode)
        assertNotNull(error.validationErrors)
        val violations: Set<ConstraintViolationEntry> = error.validationErrors!!
        assertEquals(1, violations.size)
        val violation = violations.iterator().next()
        assertEquals("must not be null", violation.message)
        assertEquals("percentage", violation.propertyPath)
    }

    /**
     * Test meaningful messages in constraints validation
     */
    @Test
    fun testBadRequestUniqueConstraint() {
        var discountCodeRes = restTemplate.postForEntity("/api/rest/discountCodes",
                DiscountCode(code = "DISCOUNT_testBadRequestUniqueConstraint", percentage = 10), DiscountCode::class.java)
        var discountCode1 = discountCodeRes.body!!
        assertEquals(201, discountCodeRes.statusCodeValue)

        val discountCode2 = discountCode1.copy(id = null)
        val errorRes = restTemplate.postForEntity("/api/rest/discountCodes", discountCode2, SimpleErrorResponse::class.java)
        assertEquals(400, errorRes.statusCodeValue)
        val error = errorRes.body!!
        assertEquals("Validation failed", error.message)
        assertEquals(400, error.httpStatusCode)
        assertNotNull(error.validationErrors)
        assertNotNull(error.validationErrors?.first())
        val violation = error.validationErrors!!.first()
        assertEquals("Unique value not available for property: code", violation.message)
        assertEquals("code", violation.propertyPath)
    }

    /**
     * Test repo's wont merge when trying to create
     */
    @Test
    fun testBadRequestUniqueId() {
        var discountCodeRes = restTemplate.postForEntity("/api/rest/discountCodes",
                DiscountCode(code = "DISCOUNT_testBadRequestUniqueId1", percentage = 6), DiscountCode::class.java)
        var discountCode1 = discountCodeRes.body!!
        assertEquals(HttpStatus.CREATED.value(), discountCodeRes.statusCodeValue)
        val discountCode2 = discountCode1.copy(code = "DISCOUNT_testBadRequestUniqueId2", percentage = 12)
        val errorRes = restTemplate.postForEntity("/api/rest/discountCodes", discountCode2, SimpleErrorResponse::class.java)
        assertEquals(HttpStatus.CONFLICT.value(), errorRes.statusCodeValue)
        val error = errorRes.body!!
        assertEquals(HttpStatus.CONFLICT.value(), error.httpStatusCode)
    }

}